pkg/audit/manager.go

```diff
@@ -185,47 +124,138 @@ func (am *auditManager) auditManagerLoop(ctx context.Context) {
+func (am *AuditManager) ensureCRDExists(ctx context.Context) error {
+	return am.client.Get(ctx, types.NamespacedName{Name: crdName}, crd)
+}
+
+func (am *AuditManager) getAllConstraintsKinds() (*metav1.APIResourceList, error) {
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(am.cfg)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	if !found {
-		return errors.Errorf("constraint %s status enforced not found", cname)
+	return discoveryClient.ServerResourcesForGroupVersion(constraintsGV)
+}
+
+func getUpdateListsFromAuditResponses(resp *constraintTypes.Responses) (map[string][]auditResult, error) {
+	updateLists := make(map[string][]auditResult)
+
+	for _, r := range resp.Results() {
+		selfLink := r.Constraint.GetSelfLink()
+		// skip if this constraint has reached the constraintViolationsLimit
+		if len(updateLists[selfLink]) == *constraintViolationsLimit {
+			continue
+		}
+
+		name := r.Constraint.GetName()
+		namespace := r.Constraint.GetNamespace()
+		apiVersion := r.Constraint.GetAPIVersion()
+		gvk := r.Constraint.GroupVersionKind()
+		message := r.Msg
+		if len(message) > msgSize {
+			message = truncateString(message, msgSize)
+		}
+		resource, ok := r.Resource.(*unstructured.Unstructured)
+		if !ok {
+			return nil, errors.Errorf("could not cast resource as reviewResource: %v", r.Resource)
+		}
+		rname := resource.GetName()
+		rkind := resource.GetKind()
+		rnamespace := resource.GetNamespace()
+		updateLists[selfLink] = append(updateLists[selfLink], auditResult{
+			cgvk:        gvk,
+			capiversion: apiVersion,
+			cname:       name,
+			cnamespace:  namespace,
+			rkind:       rkind,
+			rname:       rname,
+			rnamespace:  rnamespace,
+			message:     message,
+		})
 	}
-	if !val {
-		return errors.Errorf("constraint %s status not enforced", cname)
+	return updateLists, nil
+}
+
+func (am *AuditManager) updateConstraintsForKinds(ctx context.Context, resourceList *metav1.APIResourceList, updateLists map[string][]auditResult, timestamp string) error {
+	resourceGV := strings.Split(resourceList.GroupVersion, "/")
+	group := resourceGV[0]
+	version := resourceGV[1]
+
+	// get constraints for each Kind
+	for _, r := range resourceList.APIResources {
+		log.Info("constraint", "resource kind", r.Kind)
+		constraintGvk := schema.GroupVersionKind{
+			Group:   group,
+			Version: version,
+			Kind:    r.Kind,
+		}
+		instanceList := &unstructured.UnstructuredList{}
+		instanceList.SetGroupVersionKind(constraintGvk)
+		err := am.client.List(ctx, &client.ListOptions{}, instanceList)
+		if err != nil {
+			return err
+		}
+		log.Info("constraint", "count of constraints", len(instanceList.Items))
+		// get each constraint
+		for _, item := range instanceList.Items {
+			// if constraint is in updatedLists, then update its violations
+			constraintAuditResults, ok := updateLists[item.GetSelfLink()]
+			// else update auditTimestamp and clear violations
+			if !ok {
```

Comment:
I think we can use the `if a, b := f(); !b {}` format here. I don't think `constraintAuditResults` drops out of scope for the else {} clause, if it does we can switch the order of terms.